### ISSUE TITLE
Issues about BoundTableModel

### ISSUE THREAD

**Issue by jsfan3** [codenameone/CodenameOne#2500]
There are two bugs that reproduce always in the Simulator and in the Javascript port with a `Table `bindend to a `List<PropertyBusinessObject>`

I tried to find the cause in my code and in the Codename One sources, without success. So I created this test case.

Steps to reproduce:
- Use the following code
- Initially the table has the third column (position) with the values "0" for male, "1" for female
- Click on the column header "text_EN" two times
- Change the value of the position "male" from "0" to "2" and press Enter
- BUG: the position value of "female" changes automatically from "1" to "2"

These steps are reproduced in this video:
https://www.informatica-libera.net/videoLavoro/bugTable.mp4

Javascript version to do an immediate test:
https://cloud.codenameone.com/build/getData?m=result&i=90fdf5b2-5d28-4ea0-b3dc-4a012ae1c4fc&b=fdf966c3-373c-4995-8237-bb309a619c3c&n=preview6647905330903160885.html

Moreover, there is also another bug: the "text_EN" column is not editable and the "dbAction" is editable, but it should be the opposite, according to the following code.

The code:

```
package ...;

import static com.codename1.ui.CN.*;
import com.codename1.ui.Form;
import com.codename1.ui.Dialog;
import com.codename1.ui.plaf.UIManager;
import com.codename1.ui.util.Resources;
import com.codename1.io.Log;
import com.codename1.ui.Toolbar;
import com.codename1.ui.layouts.BoxLayout;
import com.codename1.properties.IntProperty;
import com.codename1.properties.Property;
import com.codename1.properties.PropertyBusinessObject;
import com.codename1.properties.PropertyIndex;
import com.codename1.properties.UiBinding;
import com.codename1.ui.Component;
import com.codename1.ui.table.Table;
import com.codename1.ui.table.TableLayout;
import com.codename1.ui.table.TableModel;
import java.util.ArrayList;
import java.util.List;

/**
 * TestCase for Table binded to a List<PropertyBusinessObject>
 *
 */
public class TestCase {

    private Form current;
    private Resources theme;

    private static List<PropertyBusinessObject> listToShow = null;
    private static SportTypeDAO prototype = new SportTypeDAO();

    public void init(Object context) {
        // use two network threads instead of one
        updateNetworkThreadCount(2);

        theme = UIManager.initFirstTheme("/theme");

        // Enable Toolbar on all Forms by default
        Toolbar.setGlobalToolbar(true);

        // Pro only feature
        Log.bindCrashProtection(true);

        addNetworkErrorListener(err -> {
            // prevent the event from propagating
            err.consume();
            if (err.getError() != null) {
                Log.e(err.getError());
            }
            Log.sendLogAsync();
            Dialog.show("Connection Error", "There was a networking error in the connection to " + err.getConnectionRequest().getUrl(), "OK", null);
        });
    }

    public void start() {
        if (current != null) {
            current.show();
            return;
        }
        Form hi = new Form("Test Case", BoxLayout.y());

        if (listToShow == null) {
            listToShow = getSportType();
        }

        UiBinding.BoundTableModel model = new UiBinding.BoundTableModel(listToShow, prototype);
        model.setColumnOrder(prototype.keyword, prototype.text_IT, prototype.text_EN, prototype.position, prototype.dbAction);
        model.setEditable(prototype.keyword, false);
        model.setEditable(prototype.position, true);
        model.setEditable(prototype.dbAction, false);
        model.setEditable(prototype.text_EN, true);
        model.setEditable(prototype.text_IT, true);
        Table table = new DBTable(model, 3);

        hi.add(table);

        hi.show();
    }

    public void stop() {
        current = getCurrentForm();
        if (current instanceof Dialog) {
            ((Dialog) current).dispose();
            current = getCurrentForm();
        }
    }

    public void destroy() {
    }

    /**
     * Returns sample data for this test case
     *
     * @return
     */
    private List<PropertyBusinessObject> getSportType() {
        List<PropertyBusinessObject> list = new ArrayList<>();
        SportTypeDAO row1 = new SportTypeDAO();
        SportTypeDAO row2 = new SportTypeDAO();
        row1.dbAction.set(0);
        row2.dbAction.set(0);
        row1.position.set(0);
        row2.position.set(1);
        row1.keyword.set("male");
        row2.keyword.set("female");
        row1.text_EN.set("Male");
        row2.text_EN.set("Female");
        row1.text_IT.set("Maschile");
        row2.text_IT.set("Femminile");
        list.add(row1);
        list.add(row2);
        return list;
    }

}

class SportTypeDAO implements PropertyBusinessObject {

    public final IntProperty<SportTypeDAO> dbAction = new IntProperty<>("dbAction");
    public final IntProperty<SportTypeDAO> position = new IntProperty<>("position");
    public final Property<String, SportTypeDAO> keyword = new Property<>("keyword");
    public final Property<String, SportTypeDAO> text_IT = new Property<>("text_IT");
    public final Property<String, SportTypeDAO> text_EN = new Property<>("text_EN");

    private PropertyIndex idx = new PropertyIndex(this, "SportTypeDAO",
            dbAction, position, keyword, text_IT, text_EN);

    @Override
    public PropertyIndex getPropertyIndex() {
        return idx;
    }

    public SportTypeDAO() {
    }

}

class DBTable extends Table {

    private int columnCount = 1;

    /**
     * Creates a Table useful to display a DB table
     *
     * @param model
     * @param columnToSort number of the column to sort the table (starting from
     * 0)
     *
     */
    public DBTable(TableModel model, int columnToSort) {
        super(model);
        columnCount = model.getColumnCount();
        super.setSortSupported(true);
        try {
            sort(columnToSort, false); // order by position
            Log.p("table.sort executed successfully");
        } catch (Exception ex) {
            Log.p("ERROR: cannot order table, maybe there are null values");
            Log.e(ex);
        }
    }

    @Override
    protected Component createCell(Object value, int row, int column, boolean editable) {
        Component cell = super.createCell(value, row, column, editable);
        if (row > -1) {
            cell.setUIID("TableCell");
        }
        if (row > -1 && row % 2 == 0) {
            // pinstripe effect 
            cell.getAllStyles().setBgColor(0xeeeeee);
            cell.getAllStyles().setBgTransparency(255);
        }

        return cell;
    }

    @Override
    protected TableLayout.Constraint createCellConstraint(Object value, int row, int column) {
        TableLayout.Constraint con = super.createCellConstraint(value, row, column);
        if (columnCount > 0) {
            con.setWidthPercentage(100 / columnCount);
        }
        return con;
    }
}


```

**Comment by jsfan3** [comment]
Thank you, but both issues still exist with the same test case.
I've done a pull of the Codename One project to sync it with your last commit, I compiled it, I recompiled the test case project... and nothing is changed.

**Comment by jsfan3** [comment]
About my last comment, I've done a test with the Simulator.

**Comment by codenameone** [comment]
Just tried this exact thing and it worked for me: https://www.dropbox.com/s/j1r6kz84mk9j9cc/issue-2500.mov
Can you try running in the debugger and see who invoked `setValueAt`?

**Comment by jsfan3** [comment]
Thank you for the quick reply.
I'm not sure how to do what you asked me.

Using debug, I can use the "call hierarchy".

I recorded a new video in which I show you the debugging: I see that the method model.setValueAt is called two times.

Please see and let me know if I have to do different steps for debugging:

https://www.informatica-libera.net/videoLavoro/debuggingTable.mp4

**Comment by jsfan3** [comment]
I spent the last hour asking to myself why the same steps with the same code produce different results... I hope that my last recorded video of the debug can give you an idea...

... however, can we be sure that Java acts exactly in the same way in different platforms and versions? This is not the first time that a fix that is working in yours computers doesn't work in mine (like https://github.com/codenameone/CodenameOne/issues/2493 and https://github.com/codenameone/CodenameOne/issues/2494).

So, which version of Java are you using?

I'm developing with:

java version "1.8.0_181"
Java(TM) SE Runtime Environment (build 1.8.0_181-b13)
Java HotSpot(TM) Client VM (build 25.181-b13, mixed mode, sharing)

NetBeans IDE 8.2 (Build 201609300101)

on Windows 8 32bit inside a VirtualBox machine

**Comment by jsfan3** [comment]
Good news: I discovered the cause of the different behaviour on the same code in our machines. It was really a problem of Java version. More specifically, I discovered that the Java JDK used by Netbeans wasn't the more recent one that I installed a while ago.

After reading an answer on Stack Overflow (https://stackoverflow.com/a/6958342), I modified the line 57 of the file:
`C:\Program Files\NetBeans 8.2\etc\netbeans.conf`
from:
`netbeans_jdkhome="C:\Program Files\Java\jdk1.8.0_152"`
to:
`netbeans_jdkhome="C:\Program Files\Java\jdk1.8.0_181"`
Moreover, I also deleted the directories of the older jdk installations to be sure that they are not more available.

After that, I removed all the Netbeans "dist" folders of my projects ("Codename One" and its "Java SE" port included), I recompiled "Codename One", "Java SE" and this test case project... and now I can repeat these steps:

- Start the app
- Click three times in the Left column header
- Change the first value of right from 4 to 8 and press Enter
- BUG SOLVED: the last row of right doesn't change

There is still another different bug releated to the `Table`, but I'll open a new issue.
